### PR TITLE
feat(cleanup): remove jupyterhub CR from previous release

### PR DESCRIPTION
**Pull request by zdtsw**
similar to PR https://github.com/opendatahub-io/opendatahub-operator/pull/786 but move the caller to the same one that we already have for modelmesh monitoring
some changes based on https://github.com/opendatahub-io/opendatahub-operator/pull/868

we still should get the generic cleanup function in place, but if we want to have JH CR removed in 2.8, this is a best shot for  https://issues.redhat.com/browse/RHOAIENG-443

need go along with https://github.com/opendatahub-io/odh-dashboard/pull/2513

## How Has This Been Tested?
live-build quay.io/wenzhou/opendatahub-operator-catalog:v2.8.443-20

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func main() { //nolint:funlen
 		setupLog.Error(err, "unable to update from legacy operator version")
 	}
 
-	if err = upgrade.CleanupExistingResource(setupClient, platform); err != nil {
+	if err = upgrade.CleanupExistingResource(setupClient, platform, dscApplicationsNamespace, dscMonitoringNamespace); err != nil {
```
Comment (bartoszmajsak) [inline main.go:231]: I wonder if all the `upgrade.` parts could be executed as `manager.RunnableFunc`  like the default DSCI creation  (thanks @ykaliuta for doing it). This way we would (at least) be able to propagate the controller context down the call chain.

We can go even one step further and implement `CleanupRunnable` which will implement `manager.Runnable` and `inject.Client` interface and encapsulate this functionality. This way we will have direct access to the client from the owning struct too.

The same can be done for the legacy path.

File: pkg/upgrade/upgrade.go [hunk 2]
```diff
@@ -368,41 +370,49 @@ func UpdateFromLegacyVersion(cli client.Client, platform deploy.Platform, appNS
 	return nil
 }
 
-func CleanupExistingResource(cli client.Client, platform deploy.Platform) error {
+// TODO: remove function once we have a generic solution across all components.
+func CleanupExistingResource(cli client.Client, platform deploy.Platform, dscApplicationsNamespace, dscMonitoringNamespace string) error {
 	var multiErr *multierror.Error
+	ctx := context.TODO()
```
Comment (bartoszmajsak) [inline pkg/upgrade/upgrade.go:376]: If you wrap this call in `manager.RunnableFunc`  you will have a ctx from manager, which you can pass by adding first param here.

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
I got curious about the "generic solution" for cleanup. Can you share some more details? :)

Some more ideas below :)

File: pkg/upgrade/upgrade.go [hunk 3]
```diff
@@ -368,41 +370,49 @@ func UpdateFromLegacyVersion(cli client.Client, platform deploy.Platform, appNS
...
+		multiErr = multierror.Append(multiErr, deleteDeprecatedServiceMonitors(ctx, cli, dscMonitoringNamespace, deprecatedServicemonitors))
 	}
 
+	// Handling for dashboard Jupyterhub CR, see jira #443
+	JupyterhubApp := schema.GroupVersionKind{
+		Group:   "dashboard.opendatahub.io",
+		Version: "v1",
+		Kind:    "odhapplications",
+	}
+	multiErr = multierror.Append(multiErr, removOdhApplicationsCR(ctx, cli, JupyterhubApp, "jupyterhub", dscApplicationsNamespace))
```
Comment (andrewballantyne) [inline pkg/upgrade/upgrade.go:414]: What is this deleting, file wise? I'm having a hard time reading Go here -- is https://github.com/opendatahub-io/odh-dashboard/blob/main/manifests/apps/jupyter/jupyterhub-app.yaml the resource you're deleting?

File: main.go (hunk 1 above)
Comment (zdtsw) [reply main.go:231]: just saw your comment, if i understand this, you meant the part in the webhook
yes, i agree that is a good idea to pass client around than using as param.

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (lucferbux) [reply pkg/upgrade/upgrade.go:414]: mmmm I think this is still needed, idkw it's still called jupyterhub (the file) cause the name of the odhapplication is jupyter

Comment (zdtsw) [reply pkg/upgrade/upgrade.go:414]: i do not follow this @lucferbux 
which resource you really want to delete from cluster:   the one originally generated from https://github.com/opendatahub-io/odh-dashboard/blob/main/manifests/apps/jupyter/jupyter-app.yaml?  named  "jupyter" 
or the one from  https://github.com/opendatahub-io/odh-dashboard/blob/main/manifests/apps/jupyter/jupyterhub-app.yaml, named  "juypterhub"

File: main.go [hunk 4]
```diff
@@ -228,8 +227,16 @@ func main() { //nolint:funlen
...
+	var cleanExistingResourceFunc manager.RunnableFunc = func(ctx context.Context) error {
+		if err = upgrade.CleanupExistingResource(ctx, setupClient, platform, dscApplicationsNamespace, dscMonitoringNamespace); err != nil {
+			setupLog.Error(err, "unable to perform cleanup")
+		}
+		return err
+	}
+	err = mgr.Add(cleanExistingResourceFunc)
+	if err != nil {
+		setupLog.Error(err, "error remove deprecated resources from previous version")
+		os.Exit(1)
```
Comment (VaishnaviHire) [inline main.go:239]: Should we continue with error log if there is error in resource removal i.e remove `os.Exit(1)`

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (andrewballantyne) [reply pkg/upgrade/upgrade.go:401]: This is part of https://github.com/opendatahub-io/odh-dashboard/pull/2513 -- we are not merging this sprint. So sorry @zdtsw to keep doing this to you but it has been chaotic the past few sprints. We will merge it next sprint at the start to give proper runway to test and find issues. The day before Code Freeze is a bad day to try and make this happen.

Can you linger this PR for 1 more week?

**Comment by VaishnaviHire** [comment]
/hold

Ref: https://github.com/opendatahub-io/opendatahub-operator/pull/871/files#r1507948704

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (zdtsw) [reply pkg/upgrade/upgrade.go:401]: sure, no worries. we can move it to 2.9

File: pkg/upgrade/upgrade.go [hunk 5]
```diff
@@ -555,6 +562,29 @@ func deleteDeprecatedServiceMonitors(ctx context.Context, cli client.Client, nam
 	return multiErr.ErrorOrNil()
 }
 
+func removOdhApplicationsCR(ctx context.Context, cli client.Client, gvk schema.GroupVersionKind, instanceName string, applicationNS string) error {
+	// first check if CRD in cluster
+	crd := &apiextv1.CustomResourceDefinition{}
+	if err := cli.Get(context.TODO(), client.ObjectKey{Name: fmt.Sprintf("%s.%s", gvk.Kind, gvk.Group)}, crd); err != nil {
```
Comment (bartoszmajsak) [inline pkg/upgrade/upgrade.go:568]:
```suggestion
	if err := cli.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s.%s", gvk.Kind, gvk.Group)}, crd); err != nil {
```

Comment (zdtsw) [reply pkg/upgrade/upgrade.go:568]: i must have it missed here....

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (zdtsw) [reply pkg/upgrade/upgrade.go:390]: merge this PR, since https://github.com/opendatahub-io/odh-dashboard/pull/2513 has been merged in dashboard

**Review by bartoszmajsak — APPROVED** [review]
LGTM from the code perspective. I assume that's desired removal too :)

**Comment by andrewballantyne** [comment]
> LGTM from the code perspective. I assume that's desired removal too :)

Yes, it is desirable 🙂

**Comment by andrewballantyne** [comment]
@VaishnaviHire can we get this in for 2.9? UI is merged our effort and I'm working on the ODH release now

**Comment by zdtsw** [comment]
> @VaishnaviHire can we get this in for 2.9? UI is merged our effort and I'm working on the ODH release now

dont worry, we will get it done in the nick of time

**Review by VaishnaviHire — APPROVED** [review]
/lgtm

**Pull request merged**